# Exploring eBay Car Sales Data

The aim of this project is to clean the data and analyze the used car listings.The dataset we will be using can be found [here](https://data.world/data-society/used-cars-data).

The data dictionary is as follows:
* `dateCrawled` - When this ad was first crawled. All field-values are taken from this date.
* `name` - Name of the car.
* `seller` - Whether the seller is private or a dealer.
* `offerType` - The type of listing
* `price` - The price on the ad to sell the car.
* `abtest` - Whether the listing is included in an A/B test.
* `vehicleType` - The vehicle Type.
* `yearOfRegistration` - The year in which the car was first registered.
* `gearbox` - The transmission type.
* `powerPS` - The power of the car in PS.
* `model` - The car model name.
* `kilometer` - How many kilometers the car has driven.
* `monthOfRegistration` - The month in which the car was first registered.
* `fuelType` - What type of fuel the car uses.
* `brand` - The brand of the car.
* `notRepairedDamage` - If the car has a damage which is not yet repaired.
* `dateCreated` - The date on which the eBay listing was created.
* `nrOfPictures` - The number of pictures in the ad.
* `postalCode` - The postal code for the location of the vehicle.
* `lastSeenOnline` - When the crawler saw this ad last online.

In [1]:
#Importing our Libraries
import pandas as pd
import numpy as np

#Reading our dataset from csv
autos = pd.read_csv('autos.csv', encoding='Latin-1')

Our code above did 2 things:
1. Imported libraries `pandas` and `numpy` which we will use throught out this project.
2. Extracted our dataset from `autos.csv` and assigned it to a variable named `autos`

In [2]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

Displayed above are the first 5 columns of our dataset.

In [3]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

Based on the info above we can see that our dataset has a total of 50,000 rows and 20 columns. Majority of our data are of the *object* data type, but some columns like the `yearOfRegistration` and `powerPS` are of the *int64* type.

There are also entries on specific columns that are empty. Example is `vehicleType` and `gearbox` which only have 44,905 and 47,320 entries respectively, despite there being 50,000 rows in our dataset.

Some columns as well like `offerType` and `yearOfRegistration` use [camelcase](https://en.wikipedia.org/wiki/Camel_case) in its naming convention, instead of of Python's preferred [snakecase](https://en.wikipedia.org/wiki/Snake_case).

## Cleaning Column Names:

In [4]:
#Changes the name of some columns
autos.rename(columns={'yearOfRegistration' : 'registration_year'}, inplace=True)
autos.rename(columns={'monthOfRegistration' : 'registration_month'}, inplace=True)
autos.rename(columns={'notRepairedDamage' : 'unrepaired_damage'}, inplace=True)
autos.rename(columns={'dateCreated' : 'ad_created'}, inplace=True)

#Converts the column names from camelcase to snakecase
autos.columns = autos.columns.str.lower()

Using our code above we made the following changes for easier column name readability:
* `yearOfRegistration` to `registration_year`
* `monthOfRegistration` to `registration_month`
* `notRepairedDamage` to `unrepaired_damage`
* `dateCreated` to `ad_created`
* The rest of the column names from camelcase to snakecase.

Below we can see the changes we made as well as our first 3 rows.

In [5]:
autos.head(3)

datecrawled                                        name  seller  \
0  2016-03-26 17:47:46            Peugeot_807_160_NAVTECH_ON_BOARD  privat   
1  2016-04-04 13:38:56  BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  privat   
2  2016-03-26 18:57:24                  Volkswagen_Golf_1.6_United  privat   

  offertype   price   abtest vehicletype  registration_year    gearbox  \
0   Angebot  $5,000  control         bus               2004    manuell   
1   Angebot  $8,500  control   limousine               1997  automatik   
2   Angebot  $8,990     test   limousine               2009    manuell   

   powerps   model   odometer  registration_month fueltype       brand  \
0      158  andere  150,000km                   3      lpg     peugeot   
1      286     7er  150,000km                   6   benzin         bmw   
2      102    golf   70,000km                   7   benzin  volkswagen   

  unrepaired_damage           ad_created  nrofpictures  postalcode  \
0              nein  2016-03-26 00:00:00             0       79588   
1              nein  2016-04-04 00:00:00             0       71034   
2              nein  2016-03-26 00:00:00             0       35394   

              lastseen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37

## Initial Exploration and Cleaning:

In [6]:
autos.describe(include='all')

datecrawled         name  seller offertype  price abtest  \
count                 50000        50000   50000     50000  50000  50000   
unique                48213        38754       2         2   2357      2   
top     2016-03-08 10:40:35  Ford_Fiesta  privat   Angebot     $0   test   
freq                      3           78   49999     49999   1421  25756   
mean                    NaN          NaN     NaN       NaN    NaN    NaN   
std                     NaN          NaN     NaN       NaN    NaN    NaN   
min                     NaN          NaN     NaN       NaN    NaN    NaN   
25%                     NaN          NaN     NaN       NaN    NaN    NaN   
50%                     NaN          NaN     NaN       NaN    NaN    NaN   
75%                     NaN          NaN     NaN       NaN    NaN    NaN   
max                     NaN          NaN     NaN       NaN    NaN    NaN   

       vehicletype  registration_year  gearbox       powerps  model  \
count        44905       50000.000000    47320  50000.000000  47242   
unique           8                NaN        2           NaN    245   
top      limousine                NaN  manuell           NaN   golf   
freq         12859                NaN    36993           NaN   4024   
mean           NaN        2005.073280      NaN    116.355920    NaN   
std            NaN         105.712813      NaN    209.216627    NaN   
min            NaN        1000.000000      NaN      0.000000    NaN   
25%            NaN        1999.000000      NaN     70.000000    NaN   
50%            NaN        2003.000000      NaN    105.000000    NaN   
75%            NaN        2008.000000      NaN    150.000000    NaN   
max            NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fueltype       brand unrepaired_damage  \
count       50000        50000.000000    45518       50000             40171   
unique         13                 NaN        7          40                 2   
top     150,000km                 NaN   benzin  volkswagen              nein   
freq        32424                 NaN    30107       10687             35232   
mean          NaN            5.723360      NaN         NaN               NaN   
std           NaN            3.711984      NaN         NaN               NaN   
min           NaN            0.000000      NaN         NaN               NaN   
25%           NaN            3.000000      NaN         NaN               NaN   
50%           NaN            6.000000      NaN         NaN               NaN   
75%           NaN            9.000000      NaN         NaN               NaN   
max           NaN           12.000000      NaN         NaN               NaN   

                 ad_created  nrofpictures    postalcode             lastseen  
count                 50000       50000.0  50000.000000                50000  
unique                   76           NaN           NaN                39481  
top     2016-04-03 00:00:00           NaN           NaN  2016-04-07 06:17:27  
freq                   1946           NaN           NaN                    8  
mean                    NaN           0.0  50813.627300                  NaN  
std                     NaN           0.0  25779.747957                  NaN  
min                     NaN           0.0   1067.000000                  NaN  
25%                     NaN           0.0  30451.000000                  NaN  
50%                     NaN           0.0  49577.000000                  NaN  
75%                     NaN           0.0  71540.000000                  NaN  
max                     NaN           0.0  99998.000000                  NaN

By looking at the description of our table we can see the following:
* `nrofpictures` - Only have one value which is 0. We will most likely drop this column.
* `price` - The **dollar sign** and **,** are attached to its values, we have to remove this to convert this column into numeric values.
* `odometer` - The string **km** and **,** is attached to its values, we have to remove this to convert this column into numeric values. We will also rename the column to `odometer_km`so that we don't forget that its value is in kilometers.

In [7]:
#Deletes nrofpictures column
del autos['nrofpictures']

#Deletes non_numeric characters in price
autos['price'] = autos['price'].str.replace('$' , '')
autos['price'] = autos['price'].str.replace(',' , '')
autos['price'] = autos['price'].str.replace('"' , '').astype(int)

#Deletes non_numeric characters in odometer
autos['odometer'] = autos['odometer'].str.replace('km' , '')
autos['odometer'] = autos['odometer'].str.replace(',' , '')
autos['odometer'] = autos['odometer'].str.replace('"' , '').astype(int)

#Renames odometer to odometer_km
autos.rename(columns={'odometer' : 'odometer_km'}, inplace=True)

In [8]:
autos[['price', 'odometer_km']]

price  odometer_km
0       5000       150000
1       8500       150000
2       8990        70000
3       4350        70000
4       1350       150000
...      ...          ...
49995  24900       100000
49996   1980       150000
49997  13200         5000
49998  22900        40000
49999   1250       150000

[50000 rows x 2 columns]

The table above displays the changes we made to the `price` column and our newly renamed `odometer_km` column.

## Exploring Odometer and Price Columns:

**Price:**

In [9]:
autos['price'].value_counts().sort_index(ascending=False)

99999999       1
27322222       1
12345678       3
11111111       2
10000000       1
            ... 
5              2
3              1
2              3
1            156
0           1421
Name: price, Length: 2357, dtype: int64

By looking at our data above we can see that some values in our price column goes as high as 99,999,999 and as low as 0. This is obviously inaccurate data and should be removed from our data set. The number of unique values in our `price` column is 2,357.

In [10]:
autos = autos[autos['price'].between(500, 10000000)]
autos['price'].value_counts().sort_index(ascending=False)

10000000      1
3890000       1
1300000       1
1234566       1
999999        2
           ... 
520           8
517           1
510           2
501           1
500         781
Name: price, Length: 2214, dtype: int64

By setting a realistic car price between 500 and 10,000,000 we managed to remove outliers from our dataset. Our `price` column now only has 2,214 unique values.

**Odometer:**

In [11]:
autos['odometer_km'].describe(include='all')

count     45104.000000
mean     125289.774743
std       39625.995965
min        5000.000000
25%      100000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

We can see above that the maximum value in `odometer_km` is 150,000 and 5,000 is the minimum. The mean of all the values in our column is 125,732.

In [12]:
autos['odometer_km'].unique().shape

(13,)

We only have 13 unique values in our `odometer_km` column.

In [13]:
autos['odometer_km'].value_counts().sort_index(ascending=False).head()

150000    28701
125000     4839
100000     2032
90000      1676
80000      1385
Name: odometer_km, dtype: int64

Our data above shows that majority of the values in `odometer_km` is 150,000 which has a total of 28,701 hits, followed by 125,000 which has 4,839 hits.

## Exploring the Date Columns:

In [14]:
def handle_date (date_column):
    #Removes Hours, Minutes and Seconds
    date_column = date_column.str[:10]
    
    print(round((date_column.value_counts(normalize=True, dropna=False)[:5])*100, 2))

We will use the function above to explore our columns with date values.

**Date Crawled:**

In [15]:
handle_date(autos['datecrawled'])

2016-04-03    3.88
2016-03-20    3.81
2016-03-21    3.78
2016-03-12    3.73
2016-04-04    3.66
Name: datecrawled, dtype: float64


Based on the data above, we can see that majority of our data was crawled on 2016-04-03.

**Ad Created:**

In [16]:
handle_date(autos['ad_created'])

2016-04-03    3.90
2016-03-20    3.82
2016-03-21    3.80
2016-03-12    3.71
2016-04-04    3.70
Name: ad_created, dtype: float64


Based on the data above, we can see that majority of our the Ads in our data set as well was created on 2016-04-03.

In [17]:
handle_date(autos['lastseen'])

2016-04-06    22.53
2016-04-07    13.41
2016-04-05    12.66
2016-03-17     2.77
2016-04-03     2.49
Name: lastseen, dtype: float64


Based on the data above, 22.5% of the entries our data set was last seen on 2016-04-06.

**Registration Year:**

In [18]:
autos['registration_year'].describe()

count    45104.000000
mean      2005.061879
std         89.645460
min       1000.000000
25%       2000.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

By looking at the description of our column `registration_year` above, we can see that some entries are inaccurate. An example is the minimum year found in our data set which is 1000, and the maximum is 9999.

## Dealing with Incorrect Registration Year Data:

In [19]:
#Removes rows where the registration year is not between 1886 and 2016
autos = autos[autos['registration_year'].between(1886, 2016)]

round((autos['registration_year'].value_counts(normalize=True).head(10)*100), 2)

2005    6.61
2000    6.27
2004    6.19
2006    6.16
2003    6.13
1999    5.94
2001    5.82
2002    5.58
2007    5.24
2008    5.09
Name: registration_year, dtype: float64

Using the code above, we removed the rows where the `registration_year` fell outside 1900 to 2016. We chose 2016 as our maximum value because that is when our data set was created. We then chose 1900 as our minimum value because that is the most reasonable year that is the year that cars were actually invented.

Our updated data sets shows that the most number of rows we have has the `registration_year` of 2005, where it accounts 6.6% of our total rows. Followed by the year 2000, which is 6.2% of total rows.

## Exploring Price and Mileage by Brand

**Price:**

In [20]:
#Gets only car brands that is greater than or equal to 5% of total values
greater_than_5 = (round((autos['brand'].value_counts(normalize=True)*100), 2) >= 5)
greater_than_5 = greater_than_5[greater_than_5].index
greater_than_5

Index(['volkswagen', 'bmw', 'mercedes_benz', 'opel', 'audi', 'ford'], dtype='object')

We will only be focusing on brands that have greater than or equal to 5% of total value on our data set. These brands are `volkswagen`, `bmw`, `mercedes_benz`, `opel`, `audi` and `ford`.

In [21]:
brand_mean_prices = {}

#Gets the Mean Price for our selected brands
for row in greater_than_5:
    brand_bool = autos['brand'] == row
    brand_rows = autos[brand_bool]
    
    brand_mean = brand_rows['price'].mean()
    brand_mean_prices[row] = round(brand_mean, 2)
    
brand_mean_prices

{'volkswagen': 6000.18,
 'bmw': 8828.25,
 'mercedes_benz': 8766.9,
 'opel': 3394.04,
 'audi': 9613.65,
 'ford': 4644.25}

By reading our data above, we can see that `audi`, `bmw`, and `mercedes_benz` are the most expensive, while `ford` and `opel` are the most affordable.

**Mileage:**

In [22]:
brand_mean_mileage = {}

#Gets the Mean Mileage for our selected brands
for row in greater_than_5:
    brand_bool = autos['brand'] == row
    brand_rows = autos[brand_bool]

    mean_mileage = brand_rows['odometer_km'].mean()
    brand_mean_mileage[row] = round(mean_mileage, 2)
    
brand_mean_mileage

{'volkswagen': 128239.49,
 'bmw': 132869.18,
 'mercedes_benz': 131019.19,
 'opel': 128012.42,
 'audi': 128909.72,
 'ford': 123495.04}

By reading our data above, we can see that the difference between the mean mileages of our selected brands are not far apart from one another.

## Storing Aggregate Data in a DataFrame

In [23]:
#Converts our dictionaries to Series
bmp_series = pd.Series(brand_mean_prices)
bmm_series = pd.Series(brand_mean_mileage)

#Transforms brand_mean_prices to a DataFrame
new_df = pd.DataFrame(bmp_series, columns=['mean_price'])

#Appends brand_mean_mileage to our new DataFrame
new_df['mean_mileage'] = bmm_series

new_df

mean_price  mean_mileage
volkswagen        6000.18     128239.49
bmw               8828.25     132869.18
mercedes_benz     8766.90     131019.19
opel              3394.04     128012.42
audi              9613.65     128909.72
ford              4644.25     123495.04